### ISSUE TITLE
Problem showing the route when map is loaded

### ISSUE THREAD

**Issue by hiroshi-shin** [hulop/NavCogIOSv3#55]
**Situation**: 

In the current app, it is supposed to load the map and show the route to a specific destination.

```
btnNavi.tapInside({ [weak self] _ in
    guard let _self = self else { return }
        if let n = _self.nav() {
            NavDataStore.shared().destination(byID: "EDITOR_node_1616500013756")
            let identifier = Util.routeMode == .blind ? "show_blind" : "show_general"
            n.performSegue(withIdentifier: identifier, sender: nil)
        }
    })
```

I tried to run the following hardcoded lines (copied from SearchViewController) in BlindViewController#viewDidLoad:

```
NavDataStore *nds = [NavDataStore sharedDataStore];
NavDestination *from = [NavDataStore destinationForCurrentLocation];
__block NSMutableDictionary *prefs = [@{
                        @"dist":@"500",
                        @"preset":@"9",
                        @"min_width":@"8",
                        @"slope":@"9",
                        @"road_condition":@"9",
                        @"deff_LV":@"9",
                        @"stairs":[ud boolForKey:@"route_use_stairs"]?@"9":@"1",
                        @"esc":[ud boolForKey:@"route_use_escalator"]?@"9":@"1",
                        @"elv":[ud boolForKey:@"route_use_elevator"]?@"9":@"1",
                        @"mvw":[ud boolForKey:@"route_use_moving_walkway"]?@"9":@"1",
                        @"tactile_paving":[ud boolForKey:@"route_tactile_paving"]?@"1":@"",
                        } mutableCopy];
[[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayback
                                 withOptions:AVAudioSessionCategoryOptionAllowBluetooth
                                       error:nil];
[[AVAudioSession sharedInstance] setActive:YES error:nil];
[nds requestRouteFrom:@"latlng:35.619253:139.776303:1"
                   To:@"EDITOR_node_1616500013756" withPreferences:prefs complete:^{
    __unsafe_unretained typeof(self) _self = self;
    if (_self != nil) {
        [[NSNotificationCenter defaultCenter] postNotificationName:REQUEST_PROCESS_SHOW_ROUTE object:_self];
    }
}];
```

**Problems**:

1. The `from` variable: 

- When I tried NavDataStore#destinationForCurrentLocation in `tapInside`, it gives lat and lng values of NaN.
- Inside BlindViewController#viewDidLoad, the above function gives 0.00...1 for lat and lng.

2. The `to` variable:

- Inside `tapInside`, I tried to assign a destination, but it returned `nil` in BlindViewController#viewDidLoad.

3. Even with hardcoding, the `route` in BlindViewController#requestShowRoute appears `nil`.

**Conclusion**:

Sorry for not understanding the project clearly. 
Please let me know if there is any suggestion for using `NavDataStore` correctly in this situation.
Thank you.

**Comment by daisukes** [comment]
Hi @hiroshi-shin 

Thank you for sharing the issue.
The issue was caused by a limitation of the current server-side implementation.

Since the server is stateful (not stateless or RESTful) for each user, you need to initialize the server state so that you can request to compute a route.

First, you need to load possible destinations near the user location both on local and on server-side user state.
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/SearchViewController.m#L75

You can get a notification when the destinations are ready.
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/SearchViewController.m#L65

Then you can request a route.

**Comment by hiroshi-shin** [comment]
@daisukes 

Thank you for the suggestions. I should have checked more carefully about SearchViewController.

However, when I tried to run the following code inside BlindViewController#viewDidLoad under `[self updateView];`

```
[self updateView];
[self loadDestinations:NO];
```

The first time opening the view, it's still not set, 

> (lldb) print from.singleId
> (__NSCFString *) $0 = 0x0000000281bf2670 @"latlng:0.000000:0.000000:1"
> (lldb) print to._id
> (NSString *) $1 = nil

and it gives a "No Destination" error.

I guess that I need to put it somewhere else.

In addition, I found that calling `REQUEST_PROCESS_SHOW_ROUTE` would not display the route.

I tried to copy the lines from `BlindViewController#didNavigationStarted`. It displays the route, but not moving.
Currently, I'm looking for where it really starts.

```
[NavDataStore sharedDataStore].start = [[NSDate date] timeIntervalSince1970];
dispatch_async(dispatch_get_main_queue(), ^{
    [_webView evaluateJavaScript:[NSString stringWithFormat:@"$hulop.map.getMap().getView().setZoom(%f);", [[NSUserDefaults standardUserDefaults] doubleForKey:@"zoom_for_navigation"]] completionHandler:nil];
});

NSArray *temp = [[NavDataStore sharedDataStore] route];
if (temp) {
    [[NSNotificationCenter defaultCenter] postNotificationName:REQUEST_PROCESS_SHOW_ROUTE object:self userInfo:@{@"route":temp}];
}
```

**Comment by daisukes** [comment]
@hiroshi-shin 

`NavDataStore` will load nearby destinations based on `mapCenter`.
That location is also used for `from` location. It looks like `NavDataStore` does not have it.
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/NavDataStore.m#L678-L679

`mapCenter` can be updated by location update.
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/NavDataStore.m#L461

**Comment by hiroshi-shin** [comment]
> NavDataStore will load nearby destinations based on mapCenter

Thanks for explaining. It works when I push the code inside `mapCenter` function.
The "No Destination" error is fixed after I changed some UI settings.

```
- (void) checkMapCenter:(NSTimer*)timer
{
    [_webView getCenterWithCompletion:^(HLPLocation *loc) {
        if (loc != nil) {
            ...
            [timer invalidate];
            if ([[NavDataStore sharedDataStore] reloadDestinations:NO]) {
                [NavUtil showModalWaitingWithMessage:NSLocalizedString(@"Loading, please wait",@"")];
            }
        }
    }];
}
```

```
- (void) destinationsChanged:(NSNotification*)note
{
    [NavUtil hideModalWaiting];

    NavDataStore *nds = [NavDataStore sharedDataStore];
    if ([[nds destinations] count] == 0) {
        ......
    } else {
        NavDestination *from = [NavDataStore destinationForCurrentLocation];
        NavDestination *to = [nds destinationByID:_destId];
        NSUserDefaults *ud = [NSUserDefaults standardUserDefaults];
        
        __block NSMutableDictionary *prefs = [@{
                                @"dist":@"500",
                                @"preset":@"9",
                                @"min_width":@"8",
                                @"slope":@"9",
                                @"road_condition":@"9",
                                @"deff_LV":@"9",
                                @"stairs":[ud boolForKey:@"route_use_stairs"]?@"9":@"1",
                                @"esc":[ud boolForKey:@"route_use_escalator"]?@"9":@"1",
                                @"elv":[ud boolForKey:@"route_use_elevator"]?@"9":@"1",
                                @"mvw":[ud boolForKey:@"route_use_moving_walkway"]?@"9":@"1",
                                @"tactile_paving":[ud boolForKey:@"route_tactile_paving"]?@"1":@"",
                                } mutableCopy];
        [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayback
                                         withOptions:AVAudioSessionCategoryOptionAllowBluetooth
                                               error:nil];
        [[AVAudioSession sharedInstance] setActive:YES error:nil];
        [nds requestRouteFrom:from.singleId
                           To:to._id withPreferences:prefs complete:^{
            __unsafe_unretained typeof(self) _self = self;
            if (_self != nil) {
                [_self startNavi];
            }
        }];
    }
}

- (void)startNavi {
    [NavDataStore sharedDataStore].start = [[NSDate date] timeIntervalSince1970];
    dispatch_async(dispatch_get_main_queue(), ^{
        [_webView evaluateJavaScript:[NSString stringWithFormat:@"$hulop.map.getMap().getView().setZoom(%f);", [[NSUserDefaults standardUserDefaults] doubleForKey:@"zoom_for_navigation"]] completionHandler:nil];
    });

    NSArray *temp = [[NavDataStore sharedDataStore] route];
    if (temp) {
        [[NSNotificationCenter defaultCenter] postNotificationName:REQUEST_PROCESS_SHOW_ROUTE object:self userInfo:@{@"route":temp}];
    }
}
```

In addition, sorry for not explaining clearly. Instead of only showing the route, I would like to start navigation immediately, just like "preview" and "start" function in SearchViewController.

I found that the `postLocationNotification` function is called many times as the location updated, but only once in my implementation. I would like to ask about the correct way to call it.

Thank you.

**Comment by daisukes** [comment]
Hi @hiroshi-shin 

>  Instead of only showing the route, I would like to start navigation immediately, just like "preview" and "start" function in SearchViewController.

I guess, you are trying to customize the blind navigation mode, is it right?
Basically, you don't need to explicitly call a function to start navigation, but need to set up a navigator and its delegate so that they can handle notifications properly. 

The blind navigation is managed by `NavNavigator.m` and it will show the route here.
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/blind/NavNavigator.m#L1406
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/BlindViewController.m#L1012

`BlindViewController` is the delegate class of the navigator.
https://github.com/hulop/NavCogIOSv3/blob/a25a85834f/NavCog3/BlindViewController.m#L143

The navigator observes notifications of changes in route and location to handle navigation.
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/blind/NavNavigator.m#L962
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/blind/NavNavigator.m#L1254


> I found that the postLocationNotification function is called many times as the location updated, but only once in my implementation.

I am not sure how you tested the applications, but your location manager also needs to be fixed to report the location properly.

**Comment by hiroshi-shin** [comment]
Hi, @daisukes 
First of all, thank you for explaining the delegate.
Yes. You are right. I'm trying to customize the BlindViewController, as well as the **general** one.

I though it would start navigating by using the same statements in SearchViewController#_startNavigation
https://github.com/hulop/NavCogIOSv3/blob/master/NavCogPreview/SearchViewController.m#:~:text=-%20(void)%20_startNavigation%3A(NSDictionary*)override

However, the `NavNavigator.m`'s delegate was not called.
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/BlindViewController.m#:~:text=-%20(void)didNavigationStarted%3A(NSDictionary%20*)properties

The `- (void)startNavi` function is a temporary solution for showing the route. Currently, I'm looking for the correct place which actually starts navigation.

Nonetheless, the `ViewController` seems to be implemented in a different way. Would you like to explain it as well?

Thank you.

**Comment by daisukes** [comment]
@hiroshi-shin 

The navigation is based on the localization system using BLE beacons, which calls this delegate function.
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/AppDelegate.mm#L303

You need to test the navigation functionality where BLE beacons are installed, or test with preview.
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/SearchViewController.m#L248

**Comment by hiroshi-shin** [comment]
@daisukes 

Finally, I was able to start navigation.

For the previous implementation, I didn't set `nds.previewMode` to YES.

**Previous**
```
[nds requestRouteFrom:from.singleId
                   To:to._id withPreferences:prefs complete:^{
    __unsafe_unretained typeof(self) _self = self;
    if (_self != nil) {
        [_self startNavi];
    }
}];
```

**Current**
```
[nds requestRouteFrom:from.singleId
                   To:to._id withPreferences:prefs complete:^{
    nds.previewMode = YES;
    nds.exerciseMode = NO;
}];
```

Thank you very much for explaining all those functions.

**Comment by daisukes** [comment]
@hiroshi-shin

I am glad to hear that, but I want to make sure you know how `previewMode` works. 
It's literally for previewing the route and not for the real navigation.


> Nonetheless, the ViewController seems to be implemented in a different way. Would you like to explain it as well?

Sorry, I missed this question.
As for general pedestrian mode, the navigation is handled by JavaScript.
https://github.com/hulop/MapService/blob/30863ab7822326a433cef760fbe6b7a172f0068b/MapService/WebContent/js/mobile.js#L250

**Comment by hiroshi-shin** [comment]
@daisukes 

> It's literally for previewing the route and not for the real navigation.

Actually, I saw that in SearchViewController, the Preview and Start buttons are enable in specific conditions:
https://github.com/hulop/NavCogIOSv3/blob/a25a85834ffb9644025692a69c91c26264a196ac/NavCog3/SearchViewController.m#L213-L215

It seems to be related to `isManual` property.
However, when I tried:
```
nds.previewMode = nds.isManual;
```

It always gives NO.

I will test it tomorrow with the iBeacons.

In addition, thanks for the JS link. However, I'm not sure how to programmatically set the destination (`args.navigate`) from iOS.

**Comment by hiroshi-shin** [comment]
> There are some navigation modes, normal, preview, and exercise...

I tested the app in the iBeacon area. 
I found it might be better to manually switch between preview and normal modes from my side.

> If you set dev mode, you can provide location update by dragging map

Just tried it. That's helpful outside of the iBeacon area.

> If you want to do something from native, you may need to create a function that executes javascript.

That seems to be hight customized for the current version.
I will discuss with my team leader then.

**Comment by daisukes** [comment]
@hiroshi-shin

It looks like you could solve the original issues, so I will close this issue.
Feel free to make a new issue if you get any problems/questions.
